fix: fix so that we dont blow up when encountering an "export default…

ammmze opened the pull request:
… class ..."

also adds tests to cover various scenarios

This is regarding the discussion we had over [here](https://github.com/styled-components/babel-plugin-styled-components/issues/245#issuecomment-786770782).

But the gist of the issue is we blow up when a file contains a default export of a class:

```javascript
export default class MyComponent extends React.Component {
    // ...
}
```

The error we would receive was:
```
Property declaration of ExportDefaultDeclaration expected node to be of a type ["FunctionDeclaration","TSDeclareFunction","ClassDeclaration","Expression"] but instead got "VariableDeclaration"
```

PS. I just saw your mention in #3, I'll take a deeper look at those tests as well.

ammmze committed:
fix: fix so that we dont blow up when encountering an "export default class ..."

also adds tests to cover various scenarios

ammmze commented on package.json:
```diff
@@ -14,6 +14,9 @@
     "@babel/core": "^7.5.5",
     "@babel/plugin-proposal-class-properties": "^7.5.5",
     "@babel/preset-env": "^7.5.5",
+    "babel-test": "^0.2.4",
```
I'm gonna look at the `babel-plugin-tester` they used in that other library. I chose this one because its what the style-components plugin was using. It looks like `babel-plugin-tester` has much more usage than `babel-test`, so probably better to jump over there.

ammmze replied:
I've swapped over to `babel-plugin-tester`

ammmze commented:
I tried to also add a test that would cover the `classNameMatchers` option, but in order to do so, I needed a javascript babel config instead of json, so I tried creating a `.babelrc.js` but then it failed to load the plugin because it tries to run in commonjs, so it gives me errors with the first line of `src/index.js`... `import template from '@babel/template'`. I tried changing it to `.babelrc.mjs`, but then it just stopped working. Maybe i'll have better luck via the other babel tester library.

ammmze commented on src/index.js:
```diff
@@ -41,14 +41,30 @@ export default function({ types: t }) {
-                                                ),
-                                            ])
-                                        )
+                                        const declarators = [
+                                            t.VariableDeclarator(
+                                                componentNameIdentifier,
+                                                iife.expression
+                                            ),
+                                        ]
+                                        if (
+                                            path.parentPath.isExportDefaultDeclaration()
+                                        ) {
+                                            path.parentPath.insertBefore(
+                                                t.VariableDeclaration(
+                                                    'const',
+                                                    declarators
+                                                )
+                                            )
+                                            path.replaceWith(path.node.id)
+                                        } else {
+                                            path.replaceWith(
+                                                t.VariableDeclaration(
+                                                    'const',
+                                                    declarators
+                                                )
+                                            )
+                                        }
```
I'm not very familiar with everything here, but I know enough to pretty much understand what i'm doing here. But I based this off [another project](https://github.com/rpetrich/babel-plugin-transform-async-to-promises/pull/15/files) that was facing the same issue.

ammmze commented on test/__fixtures__/wrap-class-components/output.js:
```diff
@@ -0,0 +1,28 @@
+            }
+        }
+
+        _defineProperty(_class, 'displayName', 'FancyName1')
```
hmm...i need to look at this more ... our IIFE doesn't return the `_class`

ammmze replied:
I've just pushed some updates that I think address the class wrappers. I had noticed that when we wrapped classes, the class never got returned from our IIFE. I couldn't figure out how to have it return the `_class` variable it was defining, and it felt weird having it return the class name but everything else was using this `_class` variable, so I changed it so it no longer changes the type to a `ClassExpression`. However, in doing so, we get an infinite loop, which I broke by just checking if the parent type is a `BlockExpression`.

mbrowne commented:
Thanks for the PR! I probably won't have time to do a detailed review anytime soon, so my plan is to just run the tests and also try it in a real project when the PR is ready, and if everything works then I will just merge it. I can also publish it as a public npm package.

Regarding the Jest issue with imports when you use `.babelrc.js`, I wonder if that can be addressed by adjusting the Jest configuration. Jest has experimental support for ES modules: https://jestjs.io/docs/en/next/ecmascript-modules.

ammmze committed:
fix: have wrapped classes actually return the class

ammmze commented:
> Thanks for the PR! I probably won't have time to do a detailed review anytime soon, so my plan is to just run the tests and also try it in a real project when the PR is ready, and if everything works then I will just merge it. I can also publish it as a public npm package.

Sounds like a plan! To keep things easier to maintain, I can also (in a separate PR) see about getting some github actions stuff setup to automate semver and publishing to npm.
 
> Regarding the Jest issue with imports when you use `.babelrc.js`, I wonder if that can be addressed by adjusting the Jest configuration. Jest has experimental support for ES modules: https://jestjs.io/docs/en/next/ecmascript-modules.

Thanks! I had looked at that a little bit, but ultimately I ended up not needing the test for it since we don't currently wire up the options that would use the bit of code I was attempting to test, so its basically un-used code that was copied over `isReactClass` file.

ammmze commented on test/__fixtures__/wrap-class-components/output.js:
```diff
@@ -0,0 +1,28 @@
+import React from 'react'
+export const MyComponent =
+    /*#__PURE__*/
+    (() => {
```
I wonder if we should swap the template to use `function () {\nBODY\n}` instead of an arrow function.

The way we have it currently and when I have a babel config like this:
```javascript
module.exports = {
        presets: [
            '@babel/preset-env',
            '@babel/preset-typescript',
        ],
        plugins: [
            ['styled-components', { pure: true }],
            'pure-static-props',
            '@babel/plugin-transform-runtime',
            '@babel/plugin-syntax-dynamic-import',
            '@babel/plugin-syntax-import-meta',
            '@babel/plugin-proposal-class-properties',
            '@babel/plugin-proposal-json-strings',
            '@babel/plugin-proposal-function-sent',
            '@babel/plugin-proposal-export-namespace-from',
            '@babel/plugin-proposal-numeric-separator',
            '@babel/plugin-proposal-throw-expressions',
            ['lodash', lodashOptions],
        ]
    };
```

All the arrow functions in my app get transpiled to standard functions, but the IIFE remains as an arrow function, which will break in some browsers.

mbrowne replied:
Sure, that would be fine if you want to change that. `preset-env` is supposed to take care of that, but I forget what order things are applied in...if just generating a `function` works better, you can do that.

ammmze replied:
Yea...that's what I thought too. Not sure why its not though 😕

It looks like the plugin should be running before `preset-env` based on the ordering [described by babel](https://babeljs.io/docs/en/plugins/#plugin-ordering) 🤷🏻‍♂️

ammmze replied:
But I just confirmed it with a test where I added preset-env and provided the following input:

```javascript
import React from 'react'

export const FunctionComponent = () => {
    return <div>My Component</div>
}
FunctionComponent.displayName = 'FancyName1'
FunctionComponent.defaultProps = {}
```

And here is the output:
```javascript
import React from 'react'

const FunctionComponent =
    /*#__PURE__*/
    (() => {
        const FunctionComponent = function FunctionComponent() {
            return <div>My Component</div>
        }

        FunctionComponent.displayName = 'FancyName1'
        FunctionComponent.defaultProps = {}
        return FunctionComponent
    })()

export { FunctionComponent }
```

ammmze replied:
I've been trying to see if there is a way we can get our piece to go through the presets, so it would keep it as an arrow function if someone doesn't have it configured to transform them, but I haven't found a way yet. And I wanna keep digging into the rest of our company component library packages to see what else is tripping up our tree shaking, so I'll probably just go the easy route for now and just swap the template to use a regular function.

ammmze replied:
This update has been made. We're now wrapping with standard functions instead of arrow functions.

ammmze committed:
fix: wrap in standard function instead of arrow function for better compatibility

ideally we would be able to have it go through preset-env to get it to transform the arrow functions if necessary, but we dont appear to get that for free the way we are doing this. perhaps it can be investigated in the future

mbrowne committed:
updated readme, added changelog, and prepared for publishing

mbrowne commented:
Hi, sorry for the delay. Is this ready to merge? I ran the tests and the demo works, so it looks good to me. I also updated the readme, etc.

Future PRs for CI/CD are welcome.

Thank you

ammmze commented:
> Hi, sorry for the delay. Is this ready to merge? I ran the tests and the demo works, so it looks good to me. I also updated the readme, etc.
> 
> Future PRs for CI/CD are welcome.
> 
> Thank you

👍🏻  Yep! I think this is good to go.

mbrowne commented:
Published on npm as version 0.2.0! Thanks again.

https://www.npmjs.com/package/babel-plugin-pure-static-props

CC @Mattgic @mryechkin
